### PR TITLE
[HOPSWORKS-752] Add support for running notebooks (.ipynb) in Job service

### PR DESCRIPTION
**Pull request by robzor92**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [x] Adds tests for the submitted changes (for bug fixes & features)
- [x] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by robzor92** [comment]
https://github.com/logicalclocks/hops-testing/pull/256

File: hopsworks-ear/test/spec/execution_spec.rb [hunk 1]
```diff
@@ -18,121 +18,127 @@
...
+              #get out log
+              get_execution_log(@project[:id], $job_name_4, execution_id, "out")
+              expect(json_body[:type]).to eq "OUT"
+              expect(json_body[:log]).to be_present
 
-            #get out log
-            get_execution_log(@project[:id], $job_name, execution_id, "out")
-            expect(json_body[:type]).to eq "OUT"
-            expect(json_body[:log]).to be_present
+              #wait for log aggregation
```
Comment (tkakantousis) [inline hopsworks-ear/test/spec/execution_spec.rb]: @robzor92 why `wait for log aggregation` is now after `get out log`?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/JobController.java [hunk 2]
```diff
@@ -158,8 +158,8 @@ public JobConfiguration inspectProgram(String path, Project project, Users user,
       String username = hdfsUsersBean.getHdfsUserName(project, user);
       udfso = dfs.getDfsOps(username);
       LOGGER.log(Level.INFO, "Executing job by {0} at path: {1}", new Object[]{username, path});
-      if (!path.endsWith(".jar") && !path.endsWith(".py")) {
-        throw new IllegalArgumentException("Path does not point to a jar or .py file.");
+      if (!path.endsWith(".jar") && !path.endsWith(".py") && !path.endsWith(".ipynb")) {
```
[the remark this answers is not in the document]
Comment (robzor92) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/JobController.java]: Indeed will add that

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/spark/SparkController.java [hunk 3]
```diff
@@ -187,12 +206,14 @@ private void sanityCheck(Jobs job, Users user, String path) {
     } else if (job.getJobType() != JobType.SPARK && job.getJobType() != JobType.PYSPARK) {
       throw new IllegalArgumentException(
         "Job configuration is not a Spark job configuration. Type: " + job.getJobType());
-    } else if (!Strings.isNullOrEmpty(path) && !path.endsWith(".jar") && !path.endsWith(".py")) {
-      throw new IllegalArgumentException("Path does not point to a jar or .py file.");
+    } else if (!Strings.isNullOrEmpty(path) && !path.endsWith(".jar") && !path.endsWith(".py")
```
[the remark this answers is not in the document]
Comment (robzor92) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/spark/SparkController.java]: Indeed will add it here too

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/spark/SparkController.java [hunk 4]
```diff
@@ -187,12 +206,14 @@ private void sanityCheck(Jobs job, Users user, String path) {
...
+        && !path.endsWith(".ipynb")) {
+      throw new IllegalArgumentException("Path does not point to a .jar, .py or .ipynb file.");
     }
   }
   
-  public SparkJobConfiguration inspectProgram(String path, DistributedFileSystemOps udfso) throws JobException {
+  public SparkJobConfiguration inspectProgram(String path, DistributedFileSystemOps udfso)
```
[the remark this answers is not in the document]
Comment (robzor92) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jobs/spark/SparkController.java]: I was feeling rebellious, I will revert my ways

File: hopsworks-ear/test/spec/execution_spec.rb (hunk 1 above)
Comment (robzor92) [reply hopsworks-ear/test/spec/execution_spec.rb]: There are two wait for log aggregations, one for err and one for _out_.

Comment (tkakantousis) [reply hopsworks-ear/test/spec/execution_spec.rb]: Aha ok then

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/jupyter/JupyterController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+import javax.ejb.EJB;
+import javax.ejb.Stateless;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Stateless
+public class JupyterController {
```
[the remark this answers is not in the document]
Comment (robzor92) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/jupyter/JupyterController.java]: Okay done

**Pull request merged**